pkg/pxc/encryption/keyring.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	keyring := bytes.NewBuffer(make([]byte, 0))
+	_, err = keyring.WriteString(version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write version: %v", err)
+	}
+
+	_, err = keyring.Write(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write key: %v", err)
+	}
+
+	_, err = keyring.WriteString(eof)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write eof: %v", err)
+	}
+
+	_, err = keyring.Write(sha)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write SHA sum : %v", err)
+	}
+
+	return keyring.Bytes(), nil
+}
+
+func shaSum(data []byte) ([]byte, error) {
+	h := sha256.New()
+	_, err := h.Write(data)
+	if err != nil {
+		return nil, err
+	}
+	return h.Sum(nil), nil
+}
+
+func key() ([]byte, error) {
+	keyID := keyID()
+	buf := make([]byte, 0)
+	key := bytes.NewBuffer(buf)
+
+	err := binary.Write(key, binary.LittleEndian, int64(podSize(len(keyID))))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write key pod size: %v", err)
+	}
+
+	err = binary.Write(key, binary.LittleEndian, int64(len(keyID)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write length of key id: %v", err)
+	}
+
+	err = binary.Write(key, binary.LittleEndian, int64(len(keyType)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write length of key type: %v", err)
+	}
+
+	err = binary.Write(key, binary.LittleEndian, int64(len(userID)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write length of user id: %v", err)
+	}
+
+	err = binary.Write(key, binary.LittleEndian, int64(keyLen))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write length of AES key: %v", err)
+	}
+
+	_, err = key.WriteString(keyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write key id: %v", err)
+	}
+
+	_, err = key.WriteString(keyType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write key type: %v", err)
+	}
+
+	_, err = key.WriteString(userID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write user id: %v", err)
+	}
+
+	aes, err := aesKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate AES key: %v", err)
+	}
+
+	_, err = key.Write(aes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write AES key: %v", err)
+	}
+
+	_, err = key.Write(make([]byte, paddingLen))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write padding: %v", err)
+	}
+
+	return key.Bytes(), nil
+}
+
+func aesKey() ([]byte, error) {
+	aes := make([]byte, keyLen)
+	_, err := rand.Read(aes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random sequence of bytes: %v", err)
+	}
+
+	obfuscator := []byte("*305=Ljt0*!@$Hnm(*-9-w;:")
+	i := 0
+	l := 0
+	for i < len(aes) {
+		aes[i] ^= obfuscator[l]
+		i++
+		l = (l + 1) % len(obfuscator)
+	}
+
+	return aes, nil
+}
+
+func podSize(keyIDLen int) int {
+	size := 4*8 + keyIDLen + len(keyType) + len(userID) + 8 + keyLen
+	padding := (8 - (size % 8)) % 8
+	aligned := size + padding
+	return aligned
```

Comment:
You can just return `size + padding`

Reply:
Done https://github.com/percona/percona-xtradb-cluster-operator/pull/341/commits/4ad66bcdc1b1bced828cdf15a8d4b718130dd6da